### PR TITLE
:sparkles: Upgrade to Kubernetes 1.23.3 in Quickstart and CI

### LINKED ISSUE #5995: Upgrade to Kubernetes 1.23.3 in Quickstart and CI
As noted in https://github.com/kubernetes-sigs/cluster-api/issues/5990 there is an edge case in ClusterClass patches which is affected by the CRD issue in Kubernetes 1.23.{0,1,2). I think it would be good to upgrade to that version in our quick start and in CI.

I think we should upgrade the following:

**cluster-api**:
* docs/book/src/user/quick-start.md: v1.23.0 => v1.23.3
* test/e2e/config/docker.yaml: all env vars in `variables:` which are currently set to v1.23.0 => v1.23.3
* test/infrastructure/docker/examples/*.yaml: v1.23.0 => v1.23.3
* test/framework/bootstrap/kind_provider.go: v1.23.0 => v1.23.3
* test/infrastructure/docker/internal/docker/machine.go: v1.23.0 => v1.23.3
* test/infrastructure/docker/Dockerfile: v1.23.0 => v1.23.3 (for consistency, so we can grep for v1.23.3 the next time)

**test-infra**:
* config/jobs/kubernetes-sigs/cluster-api: `INIT_WITH_KUBERNETES_VERSION`: v1.23.0 => v1.23.3

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature

### PR DESCRIPTION
**Pull request by MartinForReal**
**What this PR does / why we need it**:
Upgrade to Kubernetes 1.23.3 in Quickstart and CI

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially implements  #5995

### CODE REVIEW

**Comment by sbueringer** [comment]
@MartinForReal 
Thx, that was quick! :)

/lgtm
/assign @fabriziopandini 
TBD if we want to cherry-pick to release-1.1.

**Comment by sbueringer** [comment]
@MartinForReal 
Can you please change "Fixes #5995" to something like "Partially implements #5995". Just so that the issue is not closed automatically, as we also need a PR to test-infra to close it.

**Comment by MartinForReal** [comment]
> @MartinForReal Shangxiang Fan FTE Can you please change "Fixes #5995" to something like "Partially implements #5995". Just so that the issue is not closed automatically, as we also need a PR to test-infra to close it.

Sure. Updated

**Review by killianmuldoon — APPROVED** [review]
/lgtm

File: test/e2e/config/docker.yaml [hunk 1]
```diff
@@ -199,10 +199,10 @@ variables:
   # allowing the same e2e config file to be re-used in different Prow jobs e.g. each one with a K8s version permutation.
   # The following Kubernetes versions should be the latest versions with already published kindest/node images.
   # This avoids building node images in the default case which improves the test duration significantly.
-  KUBERNETES_VERSION_MANAGEMENT: "v1.23.0"
-  KUBERNETES_VERSION: "v1.23.0"
+  KUBERNETES_VERSION_MANAGEMENT: "v1.23.3"
+  KUBERNETES_VERSION: "v1.23.3"
   KUBERNETES_VERSION_UPGRADE_FROM: "v1.22.4"
```
Comment (fabriziopandini) [inline test/e2e/config/docker.yaml:204]: What about picking up also latest from v1.22 given that we are at it...
```suggestion
  KUBERNETES_VERSION_UPGRADE_FROM: "v1.22.6"
```

Comment (fabriziopandini) [reply test/e2e/config/docker.yaml:204]: @sbueringer ^^

**Review by fabriziopandini — COMMENTED** [review]
Great work!
only a tiny suggestion from my side

File: test/e2e/config/docker.yaml (hunk 1 above)
Comment (sbueringer) [reply test/e2e/config/docker.yaml:204]: If I see correctly there is no published image for v1.22.6 (https://hub.docker.com/r/kindest/node/tags) which means the job which uses that version has to build v1.22.6 every time.

**Comment by sbueringer** [comment]
@fabriziopandini Not sure if want to cherry-pick. I would slightly favor yes.

I think it would be good to use v1.23.3 in the  quickstart in the v1.1.0 book and in CAPD, given that the older versions have that issue with CRDs.

File: test/e2e/config/docker.yaml (hunk 1 above)
Comment (fabriziopandini) [reply test/e2e/config/docker.yaml:204]: fair point, let's hold on this

**Comment by fabriziopandini** [comment]
/approve
/lgtm

/cherry-pick v1.1
given that we are picking up the recommended version from K8s

**Comment by sbueringer** [comment]
/test pull-cluster-api-test-mink8s-main 

known flake

**Pull request merged**